### ISSUE TITLE
Ash ignores calculation's `load` when it depends on another calculation

### ISSUE THREAD

**Issue by smt116** [ash-project/ash#309]
**Describe the bug**
Ash allows specifying what a given calculation depends on. However, using `load` works only for relationships and not for other calculations. It simply ignores them (remain "not loaded").

**To Reproduce**
```elixir
calculate :remaining_funds, MoneyMap, Calculations.RemainingFunds, select: [:amount_per_buy]
calculate :next_expected_cycle_on, :date, Calculations.NextExpectedCycleOn, load: [:remaining_funds], select: [:interval, :next_expected_buy_on]
```

**Expected behavior**
I expect `remaining_funds` to be loaded within the `Calculations.NextExpectedCycleOn` calculation module, but it remains "not loaded", and I need to manually call the API's load.

** Runtime
 - Elixir version: 1.13.3
 - Erlang version: 24.2
 - OS: macOs
 - Ash version: 1.50.16

**Additional context**
https://discord.com/channels/711271361523351632/799097774523547669/963342817105707018

**Comment by sevenseacat** [comment]
Just came across this one myself (or something very similar). In my case I override the `load/3` callback in the calculation module like this:

```
  @impl true
  def load(query, _, _) do
    Ash.Query.load(query, :walk_time)
  end
```

(where `walk_time` is another calculation on the parent resource).

I can see that the load is doing something (because `walk_time` has its own set of loads that get added to the query, and I can inspect them in my `calculate/3` callback and they are present), but `walk_time` itself is not being calculated before my second calculation runs.

**Comment by zachdaniel** [comment]
Sorry, this one dropped off my radar a bit. Looking into it, it actually seems like a pretty glaring misbehavior. All that the `load/3` callback and `load` option do is ensure that the query has those items loaded, but it does not currently ensure that anything added in that way is there before `calculate/3` is called. And if that logic is added, there will also need to be logic that handles when it is given an impossible task, i.e one calculation calls another. This may require a change to the way the `load/3` callback works, for example, that may need to be `load(records, _, _)` and `MyApi.load(records, ....)` instead.

**Comment by zachdaniel** [comment]
Although, with that said, it calls for a "linearization" of the `calculate/3` functions of each calculation whereas right now the engine may do them in parallel. But if multiple calculations require a given thing to be loaded we don't want them both to load it...this may require some creative thinking.

**Comment by zachdaniel** [comment]
I may actually be able to use the engine itself to this effect as it has dependency resolution built in. Will take a stab at it.

**Comment by zachdaniel** [comment]
Okay, so its a bit difficult to test because the current stuff  "works" in order-dependent scenarios (which is why I didn't notice this before and why it doesn't break every time). I've pushed a version up to GH that should use the internal `Ash.Engine` to sort out the required dependency orders. What this will additionally mean is that it is possible to declare an order that actually isn't possible, i.e A requires B and B requires A. The error behavior will be very un-fun in those scenarios, specifically you'll get some "deadlock" error from the internal engine. But the core behavior is more important at this point. If you could try it out and let me know, that would be great :)

**Comment by sevenseacat** [comment]
Had a bit of fun upgrading my app from 1.51.2 to master to test this out (the DSL now errors on code_interface referencing a default action, and the return values for the `load/3` callback in a calculation module are different) but the short summary is that the issue is still present.

With the following calculation module:

```elixir
defmodule MyApp.ElementContext.CalculateTotalTime do
  use Ash.Calculation

  @impl true
  def load(_query, _, _) do
    [:walk_time]
  end

  @impl true
  def calculate(element_contexts, _, _) do
    # This shouldn't be necessary - the `load/3` callback above should do it - but it doesn't?
    # https://github.com/ash-project/ash/issues/309
    {:ok, element_contexts} = MyApp.Api.load(element_contexts, :walk_time)

    Enum.map(element_contexts, fn element_context ->
      IO.inspect(element_context.walk_time)
      element_context.duration_in_seconds + element_context.walk_time
    end)
  end
end
```

(Given that `walk_time` is another calculation defined on the Element resource)

Without the explicit `MyApp.Api.load` call, `element_context.walk_time` is still `#Ash.NotLoaded<:calculation>`.

**Comment by zachdaniel** [comment]
Thanks for the tip on the code_interface referencing a default action, I think that should be an easy fix.

**Comment by zachdaniel** [comment]
I'll continue looking into the issue :)

**Comment by zachdaniel** [comment]
Actually, are you sure about it erroring on referencing a default action? https://github.com/ash-project/ash/blob/main/documentation/info/upgrading.md

In the latest version of Ash (which will eventually be be 2.0.0) default actions have now changed to a more explicit pattern.

**Comment by zachdaniel** [comment]
Okay, this should now be resolved :)

Keep in mind that if you make a circular dependency, you'll get an ugly/potentially inscrutable error.

**Comment by sevenseacat** [comment]
can confirm it works 👍

**Comment by sevenseacat** [comment]
...for a single level of loading. When the calculation is nested on another load, it still doesn't seem to compute the calculation.

eg.

```elixir
defmodule MyApp.Operation.SumDurationInSeconds do
  use Ash.Calculation

  @impl true
  # segments is a has_many association on the Operation resource, duration_in_seconds is a calculated field on the Segment resource
  def load(_query, _, _), do: [segments: :duration_in_seconds]

  @impl true
  def calculate(operations, _, _) do
    # without this line, segments is loaded, but duration_in_seconds is #Ash.NotLoaded<:calculation>
    {:ok, operations} = Vexillum.Api.load(operations, segments: [:duration_in_seconds])

    Enum.map(operations, fn operation ->
      operation.segments
      |> Enum.map(& &1.duration_in_seconds)
      |> Enum.sum()
    end)
  end
end
```

**Comment by zachdaniel** [comment]
Very strange...

**Comment by zachdaniel** [comment]
I've added a test that shows this working currently. If you could reproduce this in a test that would be the fastest way for me to fix.

**Comment by joshprice** [comment]
Adding the commit for the test for reference

> I've added a test that shows this working currently. If you could reproduce this in a test that would be the fastest way for me to fix.

https://github.com/ash-project/ash/commit/6ce1b6e642060da64375f0efeb157fe5ae455dd1

**Comment by sevenseacat** [comment]
After a whole bunch of trial and error I think I've managed to narrow it down. You can kind of trick Ash into thinking it doesn't need to load the calculation, eg. on this line:

https://github.com/ash-project/ash/blob/main/test/calculation_test.exs#L196

If you change it to `|> Ash.Query.load([:best_friends_name, best_friend: [:expr_full_name]])`

Then the test fails:

```
  1) test nested calculations are loaded if necessary (Ash.Test.CalculationTest)
     test/calculation_test.exs:193
     Assertion with == failed
     code:  assert best_friends_names == [nil, "zach daniel"]
     left:  [nil, #Ash.NotLoaded<:calculation>]
     right: [nil, "zach daniel"]
     stacktrace:
       test/calculation_test.exs:201: (test)
```

In that test, it *will* calculate the `expr_full_name` of the `best_friend`, but perhaps because its already loaded the `best_friend`, the load in the calculation module doesn't think it needs to load any more?

**Comment by zachdaniel** [comment]
Great sleuthing! Turns out it was an issue with "merging" loads. When you load a relationship and add a calculation to it, you get a load data structure like this:
```elixir
%Ash.Query{load: [relationship: %Ash.Query{...., calculations: %{name: Implementation}}]}
```

and when we "merged" loads due to multiple calls to `Ash.Query.load` we were only merging the `load` statement of the respective queries. This was definitely a tough bug to find, and having a specific test case was 💯 .

Thanks so much!

**Comment by smt116** [comment]
Does it actually work?

```diff
diff --git a/test/calculation_test.exs b/test/calculation_test.exs
index bf05f79..e9b4b2c 100644
--- a/test/calculation_test.exs
+++ b/test/calculation_test.exs
@@ -53,6 +53,8 @@ defmodule Ash.Test.CalculationTest do
           constraints: [allow_empty?: true, trim?: false]
       end

+      calculate :slug, :string, expr(full_name <> "123"), load: [:full_name]
+
       calculate :expr_full_name, :string, expr(first_name <> " " <> last_name) do
         allow_async? true
       end
@@ -107,15 +109,16 @@ defmodule Ash.Test.CalculationTest do
   end

   setup do
-    User
-    |> Ash.Changeset.new(%{first_name: "zach", last_name: "daniel"})
-    |> Api.create!()
+    user =
+      User
+      |> Ash.Changeset.new(%{first_name: "zach", last_name: "daniel"})
+      |> Api.create!()

     User
     |> Ash.Changeset.new(%{first_name: "brian", last_name: "cranston"})
     |> Api.create!()

-    :ok
+    [user: user]
   end

   test "it uses default arguments" do
@@ -240,4 +243,8 @@ defmodule Ash.Test.CalculationTest do

     assert full_names == ["bob", "brian cranston", "zach daniel"]
   end
+
+  test "loads dependencies", %{user: user} do
+    assert "zach daniel123" = Api.load!(user, [:slug])
+  end
 end
```

gives

```
ash λ mix test test/calculation_test.exs
warning: use Mix.Config is deprecated. Use the Config module instead
  config/config.exs:1

..

  1) test loads dependencies (Ash.Test.CalculationTest)
     test/calculation_test.exs:247
     ** (KeyError) key :query_context not found in: %{aggregates: %{}, calculations: %{}, public?: false, resource: Ash.Test.CalculationTest.User}
     code: assert "zach daniel123" = Api.load!(user, [:slug])
     stacktrace:
       (ash 1.51.2) lib/ash/filter/filter.ex:2239: Ash.Filter.hydrate_refs/2
       (ash 1.51.2) lib/ash/filter/filter.ex:2343: anonymous fn/3 in Ash.Filter.hydrate_refs/2
       (elixir 1.13.2) lib/enum.ex:4475: Enumerable.List.reduce/3
       (elixir 1.13.2) lib/enum.ex:2442: Enum.reduce_while/3
       (ash 1.51.2) lib/ash/filter/filter.ex:2342: Ash.Filter.hydrate_refs/2
       (ash 1.51.2) lib/ash/filter/filter.ex:2094: Ash.Filter.resolve_call/2
       (ash 1.51.2) lib/ash/resource/calculation/expression.ex:117: Ash.Resource.Calculation.Expression.select/3
       (ash 1.51.2) lib/ash/query/query.ex:1022: Ash.Query.do_load/2
       (elixir 1.13.2) lib/enum.ex:2396: Enum."-reduce/3-lists^foldl/2-0-"/3
       (ash 1.51.2) lib/ash/api/api.ex:770: Ash.Api.load/4
       (ash 1.51.2) lib/ash/api/api.ex:753: Ash.Api.load/4
       (ash 1.51.2) lib/ash/api/api.ex:727: Ash.Api.load!/4
       test/calculation_test.exs:248: (test)

.......

Finished in 1.2 seconds (1.2s async, 0.00s sync)
10 tests, 1 failure
```

**Comment by zachdaniel** [comment]
This looks like a related but separate issue. Let me see what I can do.

**Comment by zachdaniel** [comment]
Okay, I've pushed a resolution for this issue, thanks for the reproduction!

**Comment by smt116** [comment]
Looks good, 👍

**Comment by joshprice** [comment]
> Okay, I've pushed a resolution for that issue, thanks for the reproduction!

https://github.com/ash-project/ash/commit/5e270288e67e12c7e28cc59982cf47492a805a87

(For reference)
